torch/testing/_internal/common_distributed.py

```diff
@@ -877,6 +888,81 @@ def _check_return_codes(self, elapsed_time) -> None:
     def is_master(self) -> bool:
         return self.rank == 0
 
+# Utility base class for distributed Multi Process Test cases
+# This abstracts the PG creation and deletion, the backends are selected based
+# on device type. The tests functions can be instantiated per device type using
+# common_device_type.instantiate_device_type_tests
+# other backends can add entry in backend() function
+class DistributedTestBase(MultiProcessTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self._spawn_processes()
+
+    def tearDown(self):
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    def backend(self, device) -> str:
+        if "cuda" in device:
+            return "nccl"
+        elif "hpu" in device :   # intel gaudi
+            return "hccl"
+        else :
+            return "gloo"
+
+    def device_handle(self, device) :
+        """
+        Get the torch module for the given device type.
+
+        This method is used to handle different device types in distributed testing.
+        It extracts the device name from the input string and returns the corresponding
+        torch module if available.
+
+        Args:
+            device (str): A string representing the device, e.g., "cuda:0", "cpu", "hpu".
+
+        Returns:
+            Union[torch.device, None]: The corresponding torch module if available,
+                                       or None if the device is not found but is an attribute of torch.
+
+        Raises:
+            RuntimeError: If the device library is not found in torch.
+
+        Example:
+            >>> self.device_handle("cuda:0")
+            <module 'torch.cuda' from '...'>
+        """
+        device_name = device.split(":")[0]
+        if hasattr(torch, device_name):
+            return getattr(torch, device_name, None)
+        else:
+            raise RuntimeError(
+                f"Device {device} library not found"
+            )
+
+    def create_pg(self, device):
+        num_visible_devices = self.device_handle(device).device_count()
+        store = torch.distributed.FileStore(self.file_name, num_visible_devices)
+        torch.distributed.init_process_group(
+            backend=self.backend(device),
+            world_size=self.world_size,
+            rank=self.rank,
+            store=store
+        )
+        if "nccl" in self.backend(device):
+            torch.cuda.set_device(self.rank)
```

Comment:
Please remove the `set_device` call before merging. We do want to make sure our library works even without user setting the device beforehand.

Reply:
Removing this causes errors on CUDA which is why I had added it.

NCCL appears to be setting multiple distributed processes to the same device. The error I am getting is this:
`Duplicate GPU detected : rank 0 and rank 1 both on CUDA device 1a000`

The same error is not seen on HPU. My understanding is this has something to do with how NCCL functions while allocating ranks without an explicit call. 

I would appreciate any insight you have on this